routes/room.go

```diff
@@ -96,40 +109,159 @@ func UpdateRoom(c *gin.Context) {
+// DisconnectUserFromRoom disconnects a user from a room in the database.
+func DisconnectUserFromRoom(c *gin.Context) {
+	var user *models.User
+	var room models.Room
+
 	ctx, cancelCtx := context.WithTimeout(c, 1000*time.Millisecond)
 	defer cancelCtx()
 
+	user, err := utilsRoutes.ExtractUserFromContext(c)
+	if err != nil {
+		log.Printf("Failed to extract user from context: %v", err)
+		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to extract user from context"})
+		return
+	}
+
 	id, err := strconv.Atoi(c.Param("id"))
 	if err != nil {
-		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid room ID"})
 		log.Printf("Failed to convert room ID: %v", err)
+		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid room ID"})
+		return
 	}
 
-	result := db.WithContext(ctx).Delete(&models.Room{}, id)
+	err = db.WithContext(ctx).First(&room, id).Error
+	if err != nil {
+		log.Printf("Failed to find document: %v", err)
+		c.JSON(http.StatusNotFound, gin.H{"error": "Room not found"})
+		return
+	}
 
-	if result.Error != nil {
-		c.JSON(http.StatusInternalServerError, gin.H{"error": "Room not deleted"})
-		log.Printf("Failed to delete document: %v", result.Error)
+	// Remove user from the connected users list of the room
+	room.UsersID, err = utils.RemoveUintFromSlice(room.UsersID, user.ID)
+	if err != nil {
+		log.Printf("User not connected to room: %v", err)
+		c.JSON(http.StatusBadRequest, gin.H{"error": "User not in room"})
 		return
-	} else if result.RowsAffected < 1 {
-		c.JSON(http.StatusNotFound, gin.H{"error": "Room not found"})
-		log.Printf("Failed to find document: %v", result.Error)
+	}
+
+	// We want to delete an empty room and keep an owner at every instant.
+	if len(room.UsersID) == 0 {
+		result := db.WithContext(ctx).Delete(&room)
+		if result.Error != nil {
+			log.Printf("Failed to delete document: %v", err)
+			c.JSON(http.StatusInternalServerError, gin.H{"error": "Room not deleted"})
+			return
+		} else if result.RowsAffected < 1 {
+			log.Printf("Failed to find document: %v", err)
+			c.JSON(http.StatusNotFound, gin.H{"error": "Room not found"})
+			return
+		}
+		log.Printf("Room %v deleted", room.ID)
+		c.JSON(http.StatusNoContent, nil)
 		return
+	} else if room.OwnerID == user.ID {
+		room.OwnerID = uint(room.UsersID[0])
 	}
 
-	c.JSON(http.StatusNoContent, nil)
+	err = db.WithContext(ctx).Save(&room).Error
+	if err != nil {
+		log.Printf("Failed to modify document: %v", err)
+		c.JSON(http.StatusInternalServerError, gin.H{"error": "Room not modified"})
+		return
+	}
+
+	// Check if user is owner to know if we give ownership to a new user
+	if room.OwnerID == user.ID {
+		if len(roomUpdate.UsersID) > 0 {
+			err = utilsRoutes.ChangeOwner(roomUpdate, db)
+			if err != nil {
+				log.Printf("Failed to change owner: %v", err)
+				c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to change owner"})
+				return
+			}
+
+		} else {
+			err = utilsRoutes.DeleteRoom(room.ID, db)
```

Comment:
Either the function isn't necessary or the error check isn't, right now it's a doublon.
Also, we know the room exist from line 201, doesn't need to check for that.